### DISCUSSION TITLE
MongoDB _id: ObjectID vs id: UUID

### DISCUSSION THREAD

**Discussion by devdupont** [fastapi-users/fastapi-users#637] [Ideas]
I've been running into some interesting integration errors while trying to use MongoDB for the user database. The core of the issue is the use of `id: UUID` in the user model for both SQL and NoSQL. I understand why it is there for SQL, but it creates duplicate IDs in Mongo since all documents are assigned `_id: bson.ObjectId` at the database level. In my opinion, this creates an anti-pattern for Mongo systems for three main reasons:

- It creates a duplicate `id` field that is only used by fastapi-users instead of standard `_id`
- It requires that an `id` field first be generated in a separate script to integrate with existing user collections
- It limits integration with other Mongo ODMs like [beanie](https://github.com/roman-right/beanie) which mock the `id` field by returning `_id`

This last point is causing some huge hang-ups because fastapi-users expects `id` to be `UUID` while beanie expects it to be `ObjectId`. When creating the various user models like `class User(models.BaseUser, Document):`, the order entered determines which side will fail.

My recommendation would be to cast the auth'd user_id values to UUID if using the SQL database or to ObjectId if using the MongoDBUserDatabase class. This logic would need to be accessible in the routers and authentication classes, so just attaching it to the user model won't work. Additionally, the user model would need to split to handle the different id field types.

I can try creating a pull request to apply this new pattern, but I figured I would start a discussion for a possible breaking change. Rather than raising errors in the actual codebase, I'd anticipate issues in other users' modeling if new users started being created without an `id: UUID` field in existing applications.

**Comment by frankie567** [comment]
Hi @flyinactor91!

Thank you for the detailed report and explanations. I completely understand the issue here.

Since we are supporting several database backends, we had to choose a common id format for every database backends. In the case of MongoDB, it indeed creates an id different than the default `_id` one.

Personnally, I don't see any bad issue with having a custom `id` field in a MongoDB database, especially with a proper index. I run several projects in production with this pattern without any issue.

The main concern is indeed with an ODM like beanie that would internally map `_id` with `id`. Clearly, I don't want to go down the path of having to handle UUID vs ObjectId everywhere in the code; with a bunch of corner cases.

However, I'm open to consider a database adapter for beanie, that could find ways to circumvent this issue. Maybe it's possible to tell beanie to NOT alias the id? Actually, I see that some improvements were made on this subject: https://github.com/roman-right/beanie/issues/12#issuecomment-853041643

Tell me what you think about it 😄
